ggshield/core/git_shell.py

```diff
@@ -154,3 +172,82 @@ def get_list_commit_SHA(
         # but returns an empty range, example git rev-list HEAD...
 
     return commit_list
+
+
+def get_filepaths_from_ref(wd: Optional[str], ref: str) -> List[str]:
+    """
+    Fetches a list of all file paths indexed at a given reference in a git repository.
+    :param wd: string path to the git repository. Defaults to current directory
+    :param ref: git reference, like a commit SHA, a relative reference like HEAD~1, ...
+    """
+    if not wd:
+        wd = os.getcwd()
+
+    check_git_ref(wd, ref)
+
+    filepaths = git(["ls-tree", "--name-only", "-r", ref], cwd=wd).splitlines()
+    return filepaths
+
+
+def get_staged_filepaths(wd: Optional[str]) -> List[str]:
+    """
+    Fetches a list of all file paths at the index in a git repository.
+    :param wd: string path to the git repository. Defaults to current directory
+    """
+    if not wd:
+        wd = os.getcwd()
+
+    filepaths = git(["ls-files", "-c"], cwd=wd).splitlines()
+    return filepaths
+
+
+def tar_from_ref_and_filepaths(
+    wd: Optional[str],
+    ref: str,
+    filepaths: List[str],
```

Comment:
nit: you can make `filepaths` an `Iterable[str]` (or even an `Iterable[Path]`, I would very much like to see more Path in the codebase 😉), this way one can use the return value of a generator as an argument.